### ISSUE TITLE
RF Selenium2Library compatibility issue with Firefox version 47.0

### ISSUE THREAD

**Issue by almahaja** [robotframework/SeleniumLibrary#614]
Robot framework Selenium2Library seems to have compatibility issue with Firefox latest version (47.0). When we try to open firefox we get a pop-up "Firefox has topped working". 
# 

Problem signature:
Problem Event Name: APPCRASH
Application Name:   firefox.exe
Application Version:    47.0.0.5999
Application Timestamp:  57535440
Fault Module Name:  xul.dll
Fault Module Version:   47.0.0.5999
Fault Module Timestamp: 575365bf
Exception Code: 80000003
Exception Offset:   00763a7f
OS Version: 6.1.7601.2.1.0.274.10
Locale ID:  1033
Additional Information 1:   0a9e
Additional Information 2:   0a9e372d3b4ad19135b953a78882e789
Additional Information 3:   0a9e
Additional Information 4:   0a9e372d3b4ad19135b953a78882e789

Read our privacy statement online:
http://go.microsoft.com/fwlink/?linkid=104288&clcid=0x0409

If the online privacy statement is not available, please read our privacy statement offline:
C:\Windows\system32\en-US\erofflps.txt
# 

I am using Robot Framewor3.0+Python 2.7.10

We tried upgrading Selenium2Library (2.53.3)

Additional Info:
- We tried Reinstalling robotframework (v2.9.2 and v3.0), also selenium2library.. But unfortunately it didn't solve the problem.
- Also tried using marionette driver using desired capabilities in our custom lib (using RF Selenium2Library). With Firefox v46 and Firefox v47, didn't find any issues for v46 but for v47 ran into same problem.

**Comment by aaltat** [comment]
We should build support for the FF 47 and marionette driver. But we must also remain compatible with older versions. Need to think how to do it

**Comment by ombre42** [comment]
I posted [a workaround on SO](http://stackoverflow.com/a/37842250/2532697).
If you follow the instructions on [Marionette](https://developer.mozilla.org/en-US/docs/Mozilla/QA/Marionette/WebDriver), you can launch the browser with Marionette enabled by doing something like this:

```
${ff default caps}    Evaluate    sys.modules['selenium.webdriver'].common.desired_capabilities.DesiredCapabilities.FIREFOX    sys,selenium.webdriver
Set To Dictionary    ${ff default caps}    marionette=${True}
Open Browser    https://www.stackoverflow.com    ff
```

For local webdrivers, it should be possible to check the Firefox version before launching it and set the capabilities appropriately. I have not looked into how this all impacts Selenium Grid.

**Comment by aaltat** [comment]
Most likely there will be some meaningful error raised by the selenium. Which can be captured and retry can be performed with the other way.

But because of this we need to do some work on the CI side which is not small effort.

**Comment by ombre42** [comment]
Perhaps. Here is what it looks like:

WebDriverException: Message: The browser appears to have exited before we could connect. If you specified a log_file in the FirefoxBinary constructor, check it for details.

20160622 15:28:17.133 : DEBUG : 
Traceback (most recent call last):
  File "<decorator-gen-26>", line 2, in open_browser
  File "C:\apps\Python27\lib\site-packages\Selenium2Library\keywords\keywordgroup.py", line 15, in _run_on_failure_decorator
    return method(_args, *_kwargs)
  File "C:\apps\Python27\lib\site-packages\Selenium2Library\keywords_browsermanagement.py", line 114, in open_browser
    browser = self._make_browser(browser_name,desired_capabilities,ff_profile_dir,remote_url)
  File "C:\apps\Python27\lib\site-packages\Selenium2Library\keywords_browsermanagement.py", line 528, in _make_browser
    browser = creation_func(remote, desired_capabilities, profile_dir)
  File "C:\apps\Python27\lib\site-packages\Selenium2Library\keywords_browsermanagement.py", line 544, in _make_ff
    browser = webdriver.Firefox(firefox_profile=profile)
  File "C:\apps\Python27\lib\site-packages\selenium\webdriver\firefox\webdriver.py", line 81, in __init__
    self.binary, timeout)
  File "C:\apps\Python27\lib\site-packages\selenium\webdriver\firefox\extension_connection.py", line 51, in **init**
    self.binary.launch_browser(self.profile, timeout=timeout)
  File "C:\apps\Python27\lib\site-packages\selenium\webdriver\firefox\firefox_binary.py", line 68, in launch_browser
    self._wait_until_connectable(timeout=timeout)
  File "C:\apps\Python27\lib\site-packages\selenium\webdriver\firefox\firefox_binary.py", line 98, in _wait_until_connectable
    raise WebDriverException("The browser appears to have exited "

**Comment by nilsfr** [comment]
I basically do the same as ombre42, but in python code. When I call the `Open Browser` keyword, a browser window is opened. After that nothing happens, the browser do not fetch the URL given to `Open Browser`. Have you managed to get Selenium to actually interact with the browser? To me it looks like Robot Framework is in some kind of dead-lock after the browser window is opened.

I use version 1.7.4 of Selenium2Library,  version 2.53.5 of Selenium and Robot Framework 3.0. 

I run on Ubuntu 16.04 with Python 2.7.11.

I do this in a Variable file written in Python:

```
# Use Marionette for Firefox
import os
from selenium.webdriver.common.desired_capabilities import DesiredCapabilities
DesiredCapabilities.FIREFOX['marionette'] = True
os.environ['PATH'] += ':' + os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', 'bin')
```

**Comment by almahaja** [comment]
Thanks Kevin, the workaround you mentioned worked for me, earlier i was
doing something different (providing marionette:true in Open Browser itself)
which was working for v46 only..

**Comment by nilsfr** [comment]
The reason that Marionette do not work for me is that the marionette driver in Firefox 47 never respond, so it is blocked on socket.recv. The geckodriver is a proxy between webdriver and marionette, so I tried to connect to marionette directly and never got any response. It looks like Marionette is broken in version 47 of Firefox.

https://github.com/mozilla/geckodriver/issues/89

**Comment by ombre42** [comment]
@nilsfr I have only attempted Firefox 47 on Windows 7. Sorry to hear its not working on Ubuntu. Wish I had some advice...

**Comment by andreabisello** [comment]
for who needs to use firefox,
a temporary solution on ubuntu can be to downgrade firefox to firefox 28 using
sudo apt-get install firefox=28.0+build2-0ubuntu2

**Comment by aaltat** [comment]
There is a new Firefox version, 47.0.1, which fixes problem with selenium. Has anyone given a try and what might be the results?

**Comment by almahaja** [comment]
For me (with Firefox v 47.0.1) its opening the Browser but fails to open
the URL and my script fail with error "No browser is open"

**Comment by aaltat** [comment]
Good to know, thanks from the input.

**Comment by aaltat** [comment]
There is indication that this is fixed with selenium 2.53.6 and Firefox 47.0.1 combination. Therefore closing the issue. 

If you feel that this is not solved, please sends a message and details why you think this is still an issue.

**Comment by almahaja** [comment]
Yes, it works only for latest selenium v2.53.6 and Firefox 47.0.1.

Thanks

**Comment by yorgabr** [comment]
I also confirm this bug.
ombre42's solution worked for me. I´m using Portable Firefox 49.0.1, RED 0.6.2.201603151414, Robot Framework 3.0, Python 2.7.11, and jPortable 1.8.0_101, under Win32 on Windows 7 64 OS.

**Comment by aaltat** [comment]
This bug and your problem are different problems. Please open a new issue about your problem.

**Comment by almahaja** [comment]
@aaltat : I am trying to use Firefox with geckodriver but i see though it works for Firefox47.02/48.0.2 but for Firefox v50.0.2 the close browser (keyword from RF Selenium2Library) fails with error pop-up "Firefox has stopped working". Is this a known problem with latest version of Firefox. Also, seems like it only works with Selenium3.0.2, am i right?

So far below are my findings:
1. RF3.0, FF47.0.2 (without geckodriver) and Selenium-2.53.6, Python 2.7.10  :- Works
2. RF3.0, FF47.0.2 + geckodriver and Selenium-3.0.2, Python 2.7.10 :- Works
3. RF3.0, FF48.0.2 + geckodriver and Selenium-3.0.2, Python 2.7.10 :- Works
4. RF3.0, FF48.0.2 + geckodriver and Selenium-2.53.6, Python 2.7.10 :- Doesn't work.
5. RF3.0, FF50.0.2 + geckodriver and Selenium-3.0.2, Python 2.7.10 :- Close Browser (Keyword from RF Selenium2Library) fails with error "Firefox has stopped working".
6. RF3.0, FF50.0.2 + geckodriver and Selenium-2.53.6, Python 2.7.10 :- Fails to open the URL and gives error ""Firefox has stopped working"."

Thanks,
-alok

**Comment by aaltat** [comment]
This is a known issue and is actually problem with selenium and Firefox. Currently selenium only supports Firefox 47.0.2 or older. Support for newer version of Firefox needs geckodriver but that needs that the selenium and Mozilla get their stuff working. It would be best to report the issues you found to the selenium or/and geckodriver issue tracker.

And yes, the current situation is very annoying.

**Comment by aaltat** [comment]
Did forget to say we also have an issue where we track the current situation, see #675.

**Comment by aaltat** [comment]
Darn a typo, did mean to say #676

**Comment by almahaja** [comment]
Thanks @aaltat i have added my findings for CLose Browser issue for v50.0.2 in https://github.com/mozilla/geckodriver/issues/285 as well.

**Comment by aaltat** [comment]
Thank you from the help. I hope that selenium and Mozilla can sort the problem soon.

**Comment by vandanna** [comment]
I use version 1.7.4 of Selenium2Library, version 2.50.0 of Selenium and Robot Framework 2.8.7

I run on Windows 7 and use Python 2.7.8

**Comment by vandanarach** [comment]
@nilsfr 
I use version 1.7.4 of Selenium2Library, version 2.50.0 of Selenium and Robot Framework 2.8.7

I run on Windows 7 and use Python 2.7.8

I am facing the below error:
Setup failed:
WebDriverException: Message: Unable to find a matching set of capabilities

Please may I know how to fix this issue?

**Comment by aaltat** [comment]
This is not a same issue. Please raise a new one or preferably ask from Robot Framework user group.
